synapse/lib/stormlib/cortex.py

```diff
@@ -0,0 +1,656 @@
+        await self.runt.snap.fire('http:resp:body', body=body)
+
+    # Convenience method
+    @s_stormtypes.stormfunc(readonly=True)
+    async def _methReply(self, code, headers=None, body=s_stormtypes.undef):
+        if self.replied:
+            raise s_exc.BadArg(mesg='Response.reply() has already been call ed.')
+
+        headers = await s_stormtypes.toprim(headers)
+        if headers is None:
+            headers = {}
+
+        if body is not s_stormtypes.undef:
+            if not isinstance(body, bytes):
+                body = await s_stormtypes.toprim(body)
+                body = json.dumps(body).encode('utf-8', 'surrogatepass')
+                headers['Content-Type'] = 'application/json'
+                headers['Content-Length'] = len(body)
+
+        await self._methSendCode(code)
+
+        if headers:
+            await self._methSendHeaders(headers)
+
+        if body is not s_stormtypes.undef:
+            await self._methSendBody(body)
+
+        self.replied = True
+        return True
+
+@s_stormtypes.registry.registerLib
+class CortexHttpApi(s_stormtypes.Lib):
+    '''
+    Library for interacting with the Extended HTTP API.
+    '''
+    _storm_locals = ()
+    _storm_lib_path = ('cortex', 'httpapi')
+
+    def getObjLocals(self):
+        return {
+            'add': self.addHttpApi,
+            'del': self.delHttpApi,
+            'get': self.getHttpApi,
+            'list': self.listHttpApis,
+            'index': self.setHttpApiIndx,
+            'response': self.makeHttpResponse,
+        }
+
+    @s_stormtypes.stormfunc(readonly=True)
+    async def makeHttpResponse(self, requestinfo):
+        requestinfo = await s_stormtypes.toprim(requestinfo)
+        return HttpReq(self.runt, requestinfo)
+
+    @s_stormtypes.stormfunc(readonly=True)
+    async def getHttpApi(self, iden):
+        s_stormtypes.confirm(('storm', 'lib', 'cortex', 'httpapi', 'get'))
+        iden = await s_stormtypes.tostr(iden)
+        adef = await self.runt.snap.core.getHttpExtApiByIden(iden)
+        return HttpApi(self.runt, adef)
+
+    @s_stormtypes.stormfunc(readonly=True)
+    async def listHttpApis(self):
+        s_stormtypes.confirm(('storm', 'lib', 'cortex', 'httpapi', 'get'))
+        adefs = await self.runt.snap.core.getHttpExtApis()
+        apis = [HttpApi(self.runt, adef) for adef in adefs]
+        return apis
+
+    async def addHttpApi(self, path, name='', desc='', runas='owner', authenticated=True, readonly=False):
+        s_stormtypes.confirm(('storm', 'lib', 'cortex', 'httpapi', 'add'))
+
+        path = await s_stormtypes.tostr(path)
+        name = await s_stormtypes.tostr(name)
+        desc = await s_stormtypes.tostr(desc)
+        runas = await s_stormtypes.tostr(runas)
+        readonly = await s_stormtypes.tobool(readonly)
+        authenticated = await s_stormtypes.tobool(authenticated)
+
+        adef = {
+            'path': path,
+            'view': self.runt.snap.view.iden,
```

Comment:
Do we want to make view and owner optional args here at creation time?

If so we'll need to updated addHttpExtApi to validate them. Currently that does no validation on those two parameters, since it is assuming it being feed a working adef.

Reply:
They can edit view after the fact and owner should always be the caller, so i think we're good as is :+1: